c++/src/kj/async-io.c++

```diff
@@ -522,7 +522,26 @@ private:
 
     void abortRead() override {
       canceler.cancel("abortRead() was called");
-      fulfiller.reject(KJ_EXCEPTION(FAILED, "read end of pipe was aborted"));
+
+      // The input might have reached EOF, but we haven't detected it yet because we haven't tried
+      // to read that far. If we had not optimized tryPumpFrom() and instead used the default
+      // pumpTo() implementation, then the input would not have called write() again once it
+      // reached EOF, and therefore the abortRead() on the other end would *not* propagate an
+      // exception! We need the same behavior here. To that end, we need to detect if we're at EOF
+      // by reading one last byte.
```

Comment:
If we're not at EOF here, and the last, one-byte read would block, wouldn't this prevent the `tryPumpFrom()` promise from fulfilling/rejecting until data becomes available?

Reply:
Yes, but that's what the generic `pumpFrom()` implementation would do as well. It would normally not notice that the output has been aborted until it gets some input which it tries to write, and that write fails.

Reply:
Ah, I see!